File: kimi-k3/reference/python/kimi_k3_ref/kda_recurrent.py
"""Naive recurrent KDA matching FLA ``naive_recurrent_kda``.

This is the clearest one-to-one mathematical form of Kimi Delta Attention.
Production inference uses chunk_kda / FlashKDA; see NUMERICS.md.
"""

from __future__ import annotations

import numpy as np


def l2_normalize(x: np.ndarray, eps: float = 1e-6) -> np.ndarray:
    """L2 normalize over the last dimension (matches use_qk_l2norm_in_kernel intent)."""
    x = np.asarray(x, dtype=np.float32)
    n = np.sqrt(np.sum(np.square(x), axis=-1, keepdims=True) + np.float32(eps))
    return (x / n).astype(np.float32)


def kda_recurrent(
    q: np.ndarray,
    k: np.ndarray,
    v: np.ndarray,
    g: np.ndarray,
    beta: np.ndarray,
    scale: float | None = None,
    initial_state: np.ndarray | None = None,
    l2norm_qk: bool = True,
) -> tuple[np.ndarray, np.ndarray]:
    """
    Parameters (FLA naive_recurrent_kda)
    ------------------------------------
    q : (B, T, H, K)
    k : (B, T, H, K)
    v : (B, T, HV, V)  — for K3, HV == H and V == K == head_dim
    g : (B, T, HV, K)  — log-space decay (output of lowerbound gate)
    beta : (B, T, HV)
    scale : default 1/sqrt(K)
    initial_state : (B, HV, K, V) or None
    l2norm_qk : if True, L2-normalize q and k (K3 kernel flag)

    Returns
    -------
    o : (B, T, HV, V)
    S : (B, HV, K, V) final state

    Recurrence (per token t, after expanding q/k to HV):
        S = S * exp(g_t)[..., None]
        delta = v_t - sum_k (k_t[..., None] * S)
        S = S + einsum(beta * k, delta)
        o_t = einsum(q_t, S)
    """
    q = np.asarray(q, dtype=np.float32)
    k = np.asarray(k, dtype=np.float32)
    v = np.asarray(v, dtype=np.float32)
    g = np.asarray(g, dtype=np.float32)
    beta = np.asarray(beta, dtype=np.float32)

    b, t, h, dk = q.shape
    hv, dv = v.shape[2], v.shape[3]
    assert hv % h == 0
    group = hv // h
    if scale is None:
        scale = float(dk) ** -0.5

    if l2norm_qk:
        q = l2_normalize(q)
        k = l2_normalize(k)

    # Expand q/k heads to value heads and scale q
    if group != 1:
        q = np.repeat(q, group, axis=2)
        k = np.repeat(k, group, axis=2)
    q = q * np.float32(scale)

    s = np.zeros((b, hv, dk, dv), dtype=np.float32)
    if initial_state is not None:
        s = s + np.asarray(initial_state, dtype=np.float32)

    o = np.zeros((b, t, hv, dv), dtype=np.float32)
    for i in range(t):
        q_i = q[:, i]  # (B, HV, K)
        k_i = k[:, i]
        v_i = v[:, i]
        g_i = g[:, i]
        b_i = beta[:, i]  # (B, HV)

        # S = S * exp(g)[..., None]
        s = s * np.exp(g_i)[..., None]
        # k·S over key dim: (B, HV, V)
        k_s = np.sum(k_i[..., None] * s, axis=-2)
        delta = v_i - k_s
        # S += (beta * k) ⊗ delta
        bk = b_i[..., None] * k_i  # (B, HV, K)
        s = s + bk[..., None] * delta[..., None, :]
        # o = q · S
        o[:, i] = np.sum(q_i[..., None] * s, axis=-2)

    return o.astype(np.float32), s.astype(np.float32)

File: kimi-k3/rust/src/kda_recurrent.rs
/// L2 normalize over last dim.
pub fn l2_normalize(x: &[f32], rows: usize, dim: usize, eps: f32, out: &mut [f32]) {
    assert_eq!(x.len(), rows * dim);
    assert_eq!(out.len(), x.len());
    for r in 0..rows {
        let mut s = 0.0f32;
        for j in 0..dim {
            let v = x[r * dim + j];
            s += v * v;
        }
        let inv = (s + eps).sqrt().recip();
        for j in 0..dim {
            out[r * dim + j] = x[r * dim + j] * inv;
        }
    }
}

/// Naive recurrent KDA (FLA `naive_recurrent_kda`) with HV == H.
pub fn kda_recurrent(
    q: &[f32],
    k: &[f32],
    v: &[f32],
    g: &[f32],
    beta: &[f32],
    batch: usize,
    time: usize,
    heads: usize,
    dim: usize,
    scale: Option<f32>,
    l2norm_qk: bool,
    out: &mut [f32],
    final_state: &mut [f32],
) {
    let scale = scale.unwrap_or((dim as f32).sqrt().recip());
    let nqk = batch * time * heads * dim;
    assert_eq!(q.len(), nqk);
    assert_eq!(k.len(), nqk);
    assert_eq!(v.len(), nqk);
    assert_eq!(g.len(), nqk);
    assert_eq!(beta.len(), batch * time * heads);
    assert_eq!(out.len(), nqk);
    assert_eq!(final_state.len(), batch * heads * dim * dim);

    let mut qq = q.to_vec();
    let mut kk = k.to_vec();
    if l2norm_qk {
        let rows = batch * time * heads;
        let mut tq = vec![0.0f32; qq.len()];
        let mut tk = vec![0.0f32; kk.len()];
        l2_normalize(&qq, rows, dim, 1e-6, &mut tq);
        l2_normalize(&kk, rows, dim, 1e-6, &mut tk);
        qq = tq;
        kk = tk;
    }
    for x in qq.iter_mut() {
        *x *= scale;
    }

    // S: (B, H, K, V) with V == dim
    let mut s = vec![0.0f32; batch * heads * dim * dim];

    for t in 0..time {
        for b in 0..batch {
            for h in 0..heads {
                let q_base = ((b * time + t) * heads + h) * dim;
                let s_base = ((b * heads + h) * dim) * dim; // (K, V) row-major: s[k*dim + v]
                let bval = beta[(b * time + t) * heads + h];

                // S *= exp(g)
                for kd in 0..dim {
                    let eg = g[q_base + kd].exp();
                    for vd in 0..dim {
                        s[s_base + kd * dim + vd] *= eg;
                    }
                }

                // k·S -> (V,)
                let mut ks = vec![0.0f32; dim];
                for vd in 0..dim {
                    let mut acc = 0.0f32;
                    for kd in 0..dim {
                        acc += kk[q_base + kd] * s[s_base + kd * dim + vd];
                    }
                    ks[vd] = acc;
                }

                // delta = v - ks
                let mut delta = vec![0.0f32; dim];
                for vd in 0..dim {
                    delta[vd] = v[q_base + vd] - ks[vd];
                }

                // S += (beta*k) ⊗ delta
                for kd in 0..dim {
                    let bk = bval * kk[q_base + kd];
                    for vd in 0..dim {
                        s[s_base + kd * dim + vd] += bk * delta[vd];
                    }
                }

                // o = q · S
                for vd in 0..dim {
                    let mut acc = 0.0f32;
                    for kd in 0..dim {
                        acc += qq[q_base + kd] * s[s_base + kd * dim + vd];
                    }
                    out[q_base + vd] = acc;
                }
            }
        }
    }
    final_state.copy_from_slice(&s);
}
